ngraph_bridge/ngraph_executor.cc

```diff
@@ -240,190 +220,167 @@ Status NGraphExecutor::GetNgExecutable(
-                           st.error_message()));
-      return errors::Internal(status_string);
-    } catch (...) {
-      BackendManager::UnlockBackend(m_op_backend_name);
-      Status st = StringToFile("tf_function_error_" + m_node_name + ".json",
-                               serialized_ng_func);
-      string status_string =
-          "Error in compiling op_backend." +
-          (st.ok() ? "" : (" Also error in dumping serialized function: " +
-                           st.error_message()));
-      return errors::Internal(status_string);
+    serialized_ng_func = itr->second;
+  }
+
+  // Serialize to nGraph if needed
+  if (std::getenv("NGRAPH_ENABLE_SERIALIZE") != nullptr) {
+    std::string file_name = "tf_function_" + m_node_name + ".json";
+    auto status_ser = StringToFile("tf_function_" + m_node_name + ".json",
+                                   serialized_ng_func);
+    if (status_ser != Status::OK()) {
+      return std::make_pair(status_ser,
+                            std::make_tuple(ng_exec, serialized_ng_func, pts));
     }
-    BackendManager::UnlockBackend(m_op_backend_name);
-    event_compile.Stop();
-    ngraph::Event::write_trace(event_compile);
-    m_ng_exec_map[signature] = ng_exec;
-
-    // caching ng_function to serialize to ngraph if needed
-    m_serialized_ng_function_map[ng_exec] = serialized_ng_func;
-
-    m_lru.push_front(signature);
-    // Memory after
-    MemoryProfile(vm, rss);
-    auto delta_vm_mem = vm - vm0;
-    auto delta_res_mem = rss - rss0;
-    NGRAPH_VLOG(1) << "NGRAPH_TF_CACHE_PROFILE: OP_ID: " << m_instance_id
-                   << " Cache length: " << m_ng_exec_map.size()
-                   << "  Cluster: " << m_node_name
-                   << " Delta VM: " << delta_vm_mem
-                   << "  Delta RSS: " << delta_res_mem
-                   << " KB Total RSS: " << rss / (1024 * 1024) << " GB "
-                   << " VM: " << vm / (1024 * 1024) << " GB" << endl;
-  }  // end of input signature not found in m_ng_exec_map
-  else {
-    // Found the input signature in m_ng_exec_map, use the cached executable
-    // Update the m_lru
-    if (signature != m_lru.front()) {
-      m_lru.remove(signature);
-      m_lru.push_front(signature);
+#if defined NGRAPH_DISTRIBUTED
+    int rank_id;
+    rank_id = ng::get_distributed_interface()->get_rank();
+    auto status = StringToFile(
+        "tf_function_" + m_node_name + "_" + to_string(rank_id) + ".json",
+        serialized_ng_func);
+    if (status != Status::OK()) {
+      return std::make_pair(status, std::make_pair(ng_exec, ng_function));
     }
-    ng_exec = it->second;
-    cache_hit = true;
-    NGRAPH_VLOG(1) << "Compilation cache hit: " << m_node_name;
+#endif
   }
-  return Status::OK();
+  // Get NgExecutable
+  auto status_ng_exec_pair =
+      GetNgExecutable(signature, ng_function, serialized_ng_func, op_backend);
+  // Create PipelinedTensorStore
+  if (status_ng_exec_pair.first == Status::OK()) {
```

Comment:
Later Refactoring: we can look at changing the signatures from tuples to just Status. That way we can use TF_RETURN_IF_ERROR and avoid the series of checking for Statuses.